src/box/iproto.cc

```diff
@@ -534,8 +542,78 @@ struct iproto_connection
 	} tx;
 	/** Authentication salt. */
 	char salt[IPROTO_SALT_SIZE];
+	/** Opened in this connection transaction or NULL */
+	struct remote_txn* txn;
+};
+
+/** Pool of remote_txn objects. */
+static struct mempool iproto_txnode_pool;
+
+/**
+ * Wrapper to store the remote transaction, queue of its messages
+ * and message to sending on disconnect.
+ */
+struct remote_txn {
+	struct txn *txn;
+	/**
+	 * List of still not processed messages for the same
+	 * transaction.
+	 */
+	struct rlist pending;
+	/*
+	 * Preallocated message to rollback the transaction in tx
+	 * thread if disconnect event occurs.
+	 * We preallocate the disconnect message at the begin of
+	 * the transaction (@sa tx_process_remote_txn()), because
+	 * we can't fail to allocate memory during connection
+	 * closing.
+	 */
+	struct iproto_msg *disconnect;
 };
 
+/**
+ * Initialize remote_txn member inside @a con
+ * @retval -1 Memory error.
+ * @retval  0 Success.
+ */
+static int
+remote_txn_create(struct iproto_connection* con)
+{
+	struct remote_txn* txn = (struct remote_txn *)
```

Comment:
why are you use mempool_alloc_xc (xc version), but not mempool_alloc? If we can't allocate memory here we ca't return NULL?

Reply:
Yes, you're right, now I looked at another allocating function at `iproto_enqueue_batch`, and made `remote_txn_create` behave similarly.